app-ng/src/handlers.lisp

```diff
@@ -0,0 +1,348 @@
+
+QVM-TOKEN is a valid persistent QVM token returned by the CREATE-QVM RPC call.
+
+SIMULATION-METHOD is a STRING naming the desired simulation method (see *AVAILABLE-SIMULATION-METHODS*).
+
+COMPILED-QUIL is a STRING containing a valid Quil program.
+
+ADDRESSES is a HASH-TABLE where the keys are mem register names and the values are either T to indicate that we should return all indices for the corresponding register, or else a LIST of the desired indices for that register.
+
+The caller must provide either QVM-TOKEN or SIMULATION-METHOD, but not both.
+
+Return a JSON-RESPONSE that contains a HASH-TABLE of the contents of the memory registers requested in the ADDRESSES request parameter."
+  (when (and (null qvm-token)
+             (not (and allocation-method simulation-method)))
+    (rpc-parameter-parse-error
+     "Either QVM-TOKEN or both ALLOCATION-METHOD and SIMULATION-METHOD must be present in the JSON request parameters."))
+  (when (and qvm-token (or allocation-method
+                           simulation-method
+                           gate-noise
+                           measurement-noise))
+    (rpc-parameter-parse-error
+     "QVM-TOKEN is incompatible with any of the following parameters: ALLOCATION-METHOD, SIMULATION-METHOD, GATE-NOISE, MEASUREMENT-NOISE."))
+  (make-json-response
+   (if qvm-token
+       (run-program-on-persistent-qvm qvm-token compiled-quil addresses)
+       (run-program-on-qvm (make-requested-qvm simulation-method
+                                               allocation-method
+                                               (quil:qubits-needed compiled-quil)
+                                               gate-noise
+                                               measurement-noise)
+                           compiled-quil
+                           addresses))))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;; ASYNC JOB HANDLERS ;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define-rpc-handler (handle-create-job "create-job") ((sub-request #'parse-sub-request))
+  "Create the request async JOB.
+
+SUB-REQUEST is a valid json object that indicates what RPC request should be run asynchronously. For example, to execute a \"run-program\" request asynchronously, you might pass the following as the SUB-REQUEST parameter:
+
+    {
+     \"type\": \"run-program\",
+     \"simulation-method\": \"pure-state\",
+     \"allocation-method\": \"native\",
+     \"num-qubits\": 10,
+     \"compiled-quil\": \"X 0\",
+     \"addresses\": {\"ro\": true}
+    }
+
+It is an error to attempt to nest \"create-job\" requests by specifying \"create-job\" as the \"type\" field of SUB-REQUEST.
+
+JOB-TOKEN is a valid JOB token returned by the CREATE-JOB RPC call.
+
+Return a JSON-RESPONSE that contains a HASH-TABLE with a \"token\" key with the newly-created JOB's unique ID token."
+  ;; TODO(appleby): should we whitelist rather than blacklist "asyncable" methods? Currently, we
```

Comment:
reminder to ponder

Reply:
Having pondered a bit, I think I might eventually switch to whitelisting once more of the API endpoints are implemented (equivalents of multishot, expectation, etc).  For now, since `run-program` is the only endpoint that does any computation, I feel like it's actually useful for testing purposes to have more than one endpoint to test with async calls, even if making an async `version` request is a little silly.